File: CSharp/BotBuilderLocation/Bing/BingGeoSpatialService.cs
namespace Microsoft.Bot.Builder.Location.Bing
{
    using System;
    using System.Configuration;
    using System.Globalization;
    using System.Linq;
    using System.Net.Http;
    using System.Threading.Tasks;
    using System.Web.Configuration;
    using Newtonsoft.Json;

    internal sealed class BingGeoSpatialService : IGeoSpatialService
    {
        private readonly static string ApiKey = WebConfigurationManager.AppSettings["BingMapsApiKey"];
        private readonly static string FindByQueryApiUrl = $"https://dev.virtualearth.net/REST/v1/Locations?key={ApiKey}&q=";
        private readonly static string FindByPointUrl = $"https://dev.virtualearth.net/REST/v1/Locations/{{0}},{{1}}?key={ApiKey}&q=";
        private readonly static string ImageUrlByPoint = $"https://dev.virtualearth.net/REST/V1/Imagery/Map/Road/{{0}},{{1}}/15?mapSize=500,500&pp={{0}},{{1}};1;{{2}}&dpi=1&logo=always&key={ApiKey}";
        private readonly static string ImageUrlByBBox = $"https://dev.virtualearth.net/REST/V1/Imagery/Map/Road?mapArea={{0}},{{1}},{{2}},{{3}}&mapSize=500,500&pp={{4}},{{5}};1;{{6}}&dpi=1&logo=always&key={ApiKey}";

        public async Task<LocationSet> GetLocationsByQueryAsync(string address)
        {
            if (string.IsNullOrEmpty(address))
            {
                throw new ArgumentNullException(nameof(address));
            }

            return await this.GetLocationsAsync(FindByQueryApiUrl + Uri.EscapeDataString(address));
        }

        public async Task<LocationSet> GetLocationsByPointAsync(double latitude, double longitude)
        {
            return await this.GetLocationsAsync(
                string.Format(CultureInfo.InvariantCulture, FindByPointUrl, latitude, longitude));
        }

        public string GetLocationMapImageUrl(Location location, int? index = null)
        {
            if (location == null)
            {
                throw new ArgumentNullException(nameof(location));
            }

            var point = location.Point;
            if (point == null)
            {
                throw new ArgumentNullException(nameof(point));
            }

            if (string.IsNullOrEmpty(ApiKey))
            {
                throw new ConfigurationErrorsException("BingMapsApiKey is missing in Web.config");
            }

            if (location.BoundaryBox != null && location.BoundaryBox.Count >= 4)
            {
                return string.Format(ImageUrlByBBox, location.BoundaryBox[0], location.BoundaryBox[1], location.BoundaryBox[2], location.BoundaryBox[3], point.Coordinates[0], point.Coordinates[1], index);
            }
            else
            {
                return string.Format(ImageUrlByPoint, point.Coordinates[0], point.Coordinates[1], index);
            }
        }

        private async Task<LocationSet> GetLocationsAsync(string url)
        {

            if (string.IsNullOrEmpty(ApiKey))
            {
                throw new ConfigurationErrorsException("BingMapsApiKey is missing in Web.config");
            }

            using (var client = new HttpClient())
            {
                var response = await client.GetStringAsync(url);
                var apiResponse = JsonConvert.DeserializeObject<LocationApiResponse>(response);

                // TODO: what is the right logic for picking a location set?
                return apiResponse.LocationSets?.FirstOrDefault();
            }
        }
    }
}

File: Node/src/services/bing-geospatial-service.ts
import * as rp from 'request-promise';
import { sprintf } from 'sprintf-js';

const findLocationByQueryUrl = "https://dev.virtualearth.net/REST/v1/Locations?";
const findLocationByPointUrl = "https://dev.virtualearth.net/REST/v1/Locations/%1$s,%2$s?";
const findImageByPointUrl = "https://dev.virtualearth.net/REST/V1/Imagery/Map/Road/%1$s,%2$s/15?mapSize=500,500&pp=%1$s,%2$s;1;%3$s&dpi=1";
const findImageByBBoxUrl = "https://dev.virtualearth.net/REST/V1/Imagery/Map/Road?mapArea=%1$s,%2$s,%3$s,%4$s&mapSize=500,500&pp=%5$s,%6$s;1;%7$s&dpi=1";

export function getLocationByQuery(apiKey: string, address: string): Promise<Array<any>> {
    var url = addKeyToUrl(findLocationByQueryUrl, apiKey) + "&q=" + encodeURIComponent(address);
    return getLocation(url);
}

export function getLocationByPoint(apiKey: string, latitude: string, longitude: string): Promise<Array<any>> {
    var url: string = sprintf(findLocationByPointUrl, latitude, longitude);
    url = addKeyToUrl(url, apiKey) + "&q=";
    return getLocation(url);
}

export function GetLocationMapImageUrl(apiKey: string, location: any, index: number) {
    if (location && location.point && location.point.coordinates && location.point.coordinates.length == 2) {

        var point = location.point;
        var url: string;

        if (location.bbox && location.bbox.length == 4) {
            url = sprintf(findImageByBBoxUrl, location.bbox[0], location.bbox[1], location.bbox[2], location.bbox[3], point.coordinates[0], point.coordinates[1], index)
        }
        else {
            url = sprintf(findImageByPointUrl, point.coordinates[0], point.coordinates[1], index)
        }            
        
        url = addKeyToUrl(url, apiKey);

        return url;
    }

    throw location;
}

function getLocation(url: string): Promise<Array<any>> {
    const requestData = {
        url: url,
        json: true
    };

    return rp(requestData)
        .then(body => {
            if (body && body.resourceSets && body.resourceSets[0] && body.resourceSets[0].resources) {
                return body.resourceSets[0].resources;
            } else {
                throw ("Invalid Api Response");
            }
        });
}

function addKeyToUrl(url: string, apiKey: string): string {
    return url + "&key=" + apiKey;
}